### ISSUE TITLE
[BUG] RunTime error with std on dataframe with bool columns 

### ISSUE THREAD

**Issue by Nanthini10** [rapidsai/cudf#3632]
**Describe the bug**

`ddf.groupby('a').b.std().compute()`

Applying the std aggregation (on a numeric column) in a dataframe having a boolean column results in the following error.

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname, udf)
    168     try:
--> 169         yield
    170     except Exception as e:

/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   4828     with raise_on_meta_error(funcname(func), udf=kwargs.pop("udf", False)):
-> 4829         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   4830 

/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/groupby.py in _var_chunk(df, *index)
    303     g = _groupby_raise_unaligned(df, by=index)
--> 304     x = g.sum()
    305 

/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py in sum(self)
     30     def sum(self):
---> 31         return self._apply_aggregation("sum")
     32 

/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py in _apply_aggregation(self, agg)
    131         """
--> 132         result = self._groupby.compute_result(agg)
    133         libcudf.nvtx.nvtx_range_pop()

/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py in compute_result(self, agg)
    369             self.sort,
--> 370             self.dropna,
    371         )

/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py in _groupby_engine(key_columns, value_columns, aggs, sort, dropna)
    550     out_key_columns, out_value_columns = libcudf.groupby.groupby(
--> 551         key_columns, value_columns, aggs, dropna=dropna
    552     )

cudf/_lib/groupby.pyx in cudf._lib.groupby.groupby()

RuntimeError: cuDF failure at: /rapids/cudf/cpp/src/groupby/common/legacy/utils.hpp:95: Invalid combination of input type and aggregation operation.

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-824de543f220> in <module>
----> 1 ddf.groupby('a').b.std().head()

/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/groupby.py in std(self, ddof, split_every, split_out)
   1358     @derived_from(pd.core.groupby.GroupBy)
   1359     def std(self, ddof=1, split_every=None, split_out=1):
-> 1360         v = self.var(ddof, split_every=split_every, split_out=split_out)
   1361         result = map_partitions(np.sqrt, v, meta=v)
   1362         return result

/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/groupby.py in var(self, ddof, split_every, split_out)
   1346             split_every=split_every,
   1347             split_out=split_out,
-> 1348             split_out_setup=split_out_on_index,
   1349         )
   1350 

/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/core.py in apply_concat_apply(args, chunk, aggregate, combine, meta, token, chunk_kwargs, aggregate_kwargs, combine_kwargs, split_every, split_out, split_out_setup, split_out_setup_kwargs, **kwargs)
   4782 
   4783     if meta is no_default:
-> 4784         meta_chunk = _emulate(chunk, *args, udf=True, **chunk_kwargs)
   4785         meta = _emulate(aggregate, _concat([meta_chunk]), udf=True, **aggregate_kwargs)
   4786     meta = make_meta(

/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   4827     """
   4828     with raise_on_meta_error(funcname(func), udf=kwargs.pop("udf", False)):
-> 4829         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   4830 
   4831 

/conda/envs/rapids/lib/python3.6/contextlib.py in __exit__(self, type, value, traceback)
     97                 value = type()
     98             try:
---> 99                 self.gen.throw(type, value, traceback)
    100             except StopIteration as exc:
    101                 # Suppress StopIteration *unless* it's the same exception that

/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname, udf)
    188         )
    189         msg = msg.format(" in `{0}`".format(funcname) if funcname else "", repr(e), tb)
--> 190         raise ValueError(msg)
    191 
    192 

ValueError: Metadata inference failed in `_var_chunk`.

You have supplied a custom function and Dask is unable to 
determine the type of output that that function returns. 

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
RuntimeError('cuDF failure at: /rapids/cudf/cpp/src/groupby/common/legacy/utils.hpp:95: Invalid combination of input type and aggregation operation.',)

Traceback:
---------
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/utils.py", line 169, in raise_on_meta_error
    yield
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/core.py", line 4829, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/conda/envs/rapids/lib/python3.6/site-packages/dask/dataframe/groupby.py", line 304, in _var_chunk
    x = g.sum()
  File "/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py", line 31, in sum
    return self._apply_aggregation("sum")
  File "/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py", line 132, in _apply_aggregation
    result = self._groupby.compute_result(agg)
  File "/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py", line 370, in compute_result
    self.dropna,
  File "/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py", line 551, in _groupby_engine
    key_columns, value_columns, aggs, dropna=dropna
  File "cudf/_lib/groupby.pyx", line 87, in cudf._lib.groupby.groupby
```

**Steps/Code to reproduce bug**

```
import cudf
import dask_cudf

df = cudf.DataFrame({'a': [1.,3,4,1.],'b': [4.,5,6,-10], 'c': [True, True, True, True]})
ddf = dask_cudf.from_cudf(df, npartitions=2)

ddf.groupby('a').b.std().compute()
```

**Expected behavior**
Compute the standard deviation on the column 'b'

**Environment overview (please complete the following information)**
 - Environment location: Docker
 - Method of cuDF install: From source 0.12

**Environment details**
```
***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=16.04
     DISTRIB_CODENAME=xenial
     DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
     NAME="Ubuntu"
     VERSION="16.04.6 LTS (Xenial Xerus)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 16.04.6 LTS"
     VERSION_ID="16.04"
     VERSION_CODENAME=xenial
     UBUNTU_CODENAME=xenial
     Linux aff06c35ee06 4.4.0-154-generic #181-Ubuntu SMP Tue Jun 25 05:29:03 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
     
     ***GPU Information***
     Tue Dec 17 23:27:36 2019       
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 410.104      Driver Version: 410.104      CUDA Version: 10.0     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |===============================+======================+======================|
     |   0  Tesla T4            Off  | 00000000:3B:00.0 Off |                    0 |
     | N/A   51C    P0    29W /  70W |   5392MiB / 15079MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   1  Tesla T4            Off  | 00000000:5E:00.0 Off |                    0 |
     | N/A   58C    P0    28W /  70W |  14379MiB / 15079MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   2  Tesla T4            Off  | 00000000:AF:00.0 Off |                    0 |
     | N/A   50C    P0    29W /  70W |  14393MiB / 15079MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   3  Tesla T4            Off  | 00000000:D8:00.0 Off |                    0 |
     | N/A   48C    P0    29W /  70W |  14403MiB / 15079MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
    
     ***CMake***
     /conda/envs/rapids/bin/cmake
     cmake version 3.16.1
     
     CMake suite maintained and supported by Kitware (kitware.com/cmake).
     
     ***g++***
     /usr/bin/g++
     g++ (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
     Copyright (C) 2015 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
     
     
     ***nvcc***
     /usr/local/cuda/bin/nvcc
     nvcc: NVIDIA (R) Cuda compiler driver
     Copyright (c) 2005-2018 NVIDIA Corporation
     Built on Sat_Aug_25_21:08:01_CDT_2018
     Cuda compilation tools, release 10.0, V10.0.130
     
     ***Python***
     /conda/envs/rapids/bin/python
     Python 3.6.7
    
```

**Comment by jrhemstad** [comment]
@devavret does the STD aggregation use the legacy `target_type_t` infrastructure?

This exception is getting hit here: https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/src/groupby/common/legacy/utils.hpp#L93

Which is inside the `target_dtypes` function that determines the output type for the combination of (input type, aggregation). It uses the machinery `target_type` traits here: https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/src/groupby/common/legacy/type_info.hpp#L55

It requires having a partial specialization for every operator, but there isn't one for STD. I assumed this is because this machinery isn't used for STD, but maybe I am mistaken.

**Comment by kkraus14** [comment]
@Nanthini10 does the issue reproduce in using just cudf without Dask? It looks like it's triggering on a `sum` aggregation.

**Comment by ayushdg** [comment]
FWIW dask's standard deviation does not invoke cudf's standard deviation aggregation at all. The core logic for `standard deviation/ variance` is implemented here: https://github.com/dask/dask/blob/master/dask/dataframe/groupby.py#L296-L335

**Comment by Nanthini10** [comment]
@kkraus14 cudf doesn't yet support std aggregation on SeriesGroupby. More context here: https://github.com/rapidsai/cudf/issues/3429 

And using the entire dataframe for the aggregation results in the following error (this is expected) - because it's trying to compute std on the boolean column too 
```
import cudf
import dask_cudf

df = cudf.DataFrame({'a': [1.,3,4,1.],'b': [4.,5,6,-10], 'c': [True, True, True, True]})
df.groupby('a').std().compute()
```
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-10-47de0950764d> in <module>
      3 
      4 df = cudf.DataFrame({'a': [1.,3,4,1.],'b': [4.,5,6,-10], 'c': [True, True, True, True]})
----> 5 df.groupby('a').std().compute()

/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/groupby/groupby.py in std(self, ddof)
    207 
    208         out_key_columns, out_value_columns = libcudf.reduce.group_std(
--> 209             key_cols, val_cols, ddof
    210         )
    211         return self._groupby.construct_result(

cudf/_lib/reduce.pyx in cudf._lib.reduce.group_std()

RuntimeError: cuDF failure at: /rapids/cudf/cpp/src/reductions/legacy/group_std.cu:116: Only numeric types are supported in variance
```

**Comment by devavret** [comment]
> does the STD aggregation use the legacy `target_type_t` infrastructure?

It doesn't. The result type was hardcoded to be `double`. 

> It looks like it's triggering on a `sum` aggregation.

That's right. This issue seems to be that the legacy `target_type_t` machinery specifies the result type for integers and floats only and not for cudf::bool8. cuDF's standard deviation does not use the sum aggregation at all.

> FWIW dask's standard deviation does not invoke cudf's standard deviation aggregation at all.

Oh so this is why calling std from dask calls sum in cudf. Because it bypasses cudf's std.

**Comment by kkraus14** [comment]
Blocked by #3581

**Comment by Nanthini10** [comment]
Closed by #3581
